gdal/frmts/hdf5/bagdataset.cpp

```diff
@@ -3375,15 +3373,13 @@ void BAGDataset::LoadMetadata()
             }
 
             adfGeoTransform[0] = dfLLX;
-            adfGeoTransform[1] = (dfURX - dfLLX) / (m_nLowResWidth - (bPixelCenterConvention ? 1 : 0));
-            adfGeoTransform[3] = dfURY;
-            adfGeoTransform[5] = (dfLLY - dfURY) / (m_nLowResHeight - (bPixelCenterConvention ? 1 : 0));
+            adfGeoTransform[1] = dfResWidth;
```

Comment:
shoud be modiied to behave nicely when PixelIsArea convention is detected

Reply:
Our understanding of the previous code was that PixelIsArea was assumed when the upper left corner was not written correctly.  This was resulting in the resolution being scaled, which is not what was intended.  As was described in #1643, the BAG is defined by the lower lefter corner and the resolution.  The upper right corner is only written because of our understanding of  ISO 19115, and apparently is being written incorrectly because there is no check on its validity when written by the associated software.  We hope to correct this.

Reply:
> Our understanding of the previous code was that PixelIsArea was assumed when the upper left corner was not written correctly. This was resulting in the resolution being scaled, which is not what was intended.

your proposed changes regarding set adfGeoTransform[] look good, except than the half-pixel shift of the origin must not be done in the PixelIsArea/PixelCorner case.

Reply:
We did the half shift to the extracted GeoTransform because it was consistent with the rest of the GDAL BAG driver.  It looked like the way the driver was written (and all the tests worked) took care of the shift during reading and writing, thus the GDAL BAG object took on a PixelIsCorner convention.  Should it not be that way?  Should we should modify the GeoTransform (and all the tests) to keep the BAG as PixelIsArea as a GDAL object? There must be an attribute we would set that informs a transform to or from other object type of any conversion that needs to be made to the associated GeoTransform.

Reply:
> We did the half shift to the extracted GeoTransform because it was consistent with the rest of the GDAL BAG driver

That's fine. I meant that when the convention of the product is detected to be the non-standard/erroneous PixelCorner/PixelIsArea then this half-pixel shift is not needed, since it is already in the georeferencing convention used by GDAL.

Reply:
From our perspective the non-standard detection indicates that the upper right corner was written incorrectly and thus a warning should be thrown.  It does not mean the BAG is PixelIsCorner.  Since the code (now) always recalculates the upper right corner there is no need to make the half pixel shift conditional.  The resolution and lower left corner are always correct and the raster is always PixelIsArea.

Reply:
>  It does not mean the BAG is PixelIsCorner

Hum, perhaps... Is there a way of checking what the correct georeferencing would be for a product like
https://data.ngdc.noaa.gov/platforms/ocean/nos/coast/H12001-H14000/H12525/BAG/H12525_MB_4m_MLLW_1of2.bag which was referenced in #1643 ? If indeed this is just an issue with the computation of the upper-right corner, then https://github.com/OSGeo/gdal/pull/1728/commits/d455285c787046d05e1215496441f7248bc4094a should be reverted. This is becoming confusing... :-)

Reply:
We went back and compared the source raster with the derived BAG for a few of our products and found the lower left corner and resolution were correct.  The upper right corner was reported as being one grid resolution away from what is reported in the BAG, so clearly it was being ignored by the software.  We think this confirms the steps taken here.
We often say these things are not hard, but there are confusing.  Thanks for accepting the work.